proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,270 @@
+       HLSLRootSignature 'main.RS'
+
+```
+
+* LLVM IR for the root signature saved in ConstantStruct.
+
+```
+
+  ; named metadata for entry root signature
+  !hlsl.entry.rootsignatures = !{!2}
+  …
+  ; link the global variable to entry function
+  !2 = !{ptr @main,
+    %0 {
+       %"dxbc::RootSignature::ContainerRootSignatureDesc" 
+          { i32 2, i32 0, i32 24, i32 0, i32 96, i32 3 }, 
+      [6 x %"dxbc::RootSignature::ContainerRootParameter"]  
+        [%"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 2, i32 0, i32 200 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 3, i32 0, i32 212 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 4, i32 0, i32 224 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 0, i32 0, i32 236 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 0, i32 0, i32 316 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 1, i32 0, i32 348 }], 
+      [2 x %"dxbc::RootSignature::StaticSamplerDesc"]
+         [%"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 85, i32 1, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+             i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 1, 
+             i32 0, i32 0 }, 
+          %"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 21, i32 3, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+           i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 2, 
+           i32 0, i32 0 }], 
+      %1 { 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          { i32 0, i32 1, i32 8 }, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %2 { %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+        { i32 3, i32 244 }, 
+        [3 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+         [%"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 2, i32 1, i32 1, i32 0, i32 0, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 0, i32 8, i32 1, i32 0, i32 1, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 1, i32 -1, i32 1, i32 0, i32 1, i32 -1 }] 
+       }, 
+       %3 {
+           %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+           { i32 1, i32 324 }, 
+           [1 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+           [%"dxbc::RootSignature::ContainerDescriptorRange" 
+             { i32 3, i32 4, i32 0, i32 1, i32 0, i32 -1 }] }, 
+           %"dxbc::RootSignature::ContainerRootConstants" 
+             { i32 10, i32 0, i32 3 } 
+       }
+      }
+    }
+
+```
+
+* LLVM IR for root signature saved in metadata.
+
+```
+
+  ; named metadata for entry root signature
+  !hlsl.entry.rootsignatures = !{!3}
+  …
+  !3 = !{ptr @main, !4}
```

Comment:
Thanks, seeing it like this helps me understand what's being talked about a bit more!

I think the schema you've followed here is to have a 1:1 mapping between the d3d struct and the metadata tuple?  I'm trying to work out what the contents of `!4` example corresponds to.  It has this data in it:

* 2
* 0
* 24
* 0
* 96
* 3
* !5, !12, !15

The RootFlags in your example is `ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT | DENY_VERTEX_SHADER_ROOT_ACCESS` which is `0x1 | 0x2` which is 3.  So I guess that the corresponds to the "3".

The example has 6 root parameters and 2 static samplers.  I'm not sure how this corresponds to the other numbers (2, 0, 24, 0, and 96).  The `!5` looks like it references metadata listing the parameters, the `!12` looks like it references a list of statics samplers.  I'm not sure what `!15` is.

It'd be great to understand what the schema is here.  Maybe the example here isn't meant to encode the root signature example above?



When coming up with schemas it would be good to consider what's involved in building up the root signature it describes, and what would be involved in lowering it to DXIL / SPIR-V target.  For example, could parameters be added by just appending some metadata and adding a reference to it to an existing list? Or would we need to chase around and update multiple tuples?  I think that the main consideration here is for code complexity rather than performance.

Reply:
It is mapping struct like
```

  struct SerializedRS {
    ContainerRootSignatureDesc RSDesc;
    ContainerRootParameter  rootParameters[6];
    StaticSamplerDesc  samplers[2];

    // Extra part for each RootParameter in rootParameters.
    // RootConstants/RootDescriptorTable/RootDescriptor dependent on ParameterType.
    // For RootDescriptorTable, the extra part will be like
    // struct {
    //   RootDescriptorTable table;
    //   ContainerDescriptorRange ranges[NumDescriptorRangesForTheTable];
    // };

    struct {

      RootConstants b0;

      ContainerRootDescriptor t1;

      ContainerRootDescriptor u1;

      struct {

        ContainerRootDescriptor tab0;

        ContainerDescriptorRange tab0Ranges[2];

      } table0;

      struct {

        ContainerRootDescriptor tab1;

        ContainerDescriptorRange tab1Ranges[1];

      } table1;

      RootConstants b10;

    };

  };

```

Reply:
For code complexity, we could take DxilMDHelper::EmitSignatureElement  in https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/DXIL/DxilMetadataHelper.cpp#L541 as example for write a struct to metadata.

To append more things, we used this tag thing in dxc https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/DXIL/DxilMetadataHelper.cpp#L1290
Not sure if it is the correct llvm approach though.

Reply:
Why would we include things like "RootParametersOffset" in the metadata?  That's the offset in bytes to the root parameters inside a binary serialization of the root signature?  I'm not sure that makes sense here.  So in the above example, for the metadata we have:

* 2: the version of the binary serialized root signature (this isn't the binary serialized root sig, so isn't relevant)
* 0: the number of parameters?? (but this should be 6?)
* 24: the offset in the binary serialized root signature (doesn't make sense here)
* 0: the number of static samplers (but this should be 2?)
* 96: another offset in the binary serialized root signature
* 3: the flags
* !5, !12, !15

I still don't understand what the !15 is for.

I think that trying to encode the binary serialized form of the root signature in metadata is missing the point of why we'd be using the metadata form in the first place.  

At the very least, if you strip out these unnecessary bits you end up with something more like:

```
!4 = !{ i32 3, !5, !12 }  # flags, parameters, static samplers
```

But I think that once you embrace that the metadata schema isn't just a transliteration of the serialized binary format into metadata nodes there's opportunity for further simplifications.

Reply:
Updated the metadata.
All unused data is removed.

Reply:
Thanks, this is looking better now.